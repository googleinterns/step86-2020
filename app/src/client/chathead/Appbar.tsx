import React, { ReactChildren, ReactElement } from "react";

import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import PictureInPictureIcon from '@material-ui/icons/PictureInPicture';
import SettingsOverscanIcon from '@material-ui/icons/SettingsOverscan';
import PhotoSizeSelectSmallIcon from '@material-ui/icons/PhotoSizeSelectSmall';
import { Toolbar, Typography, IconButton, AppBar, Box } from "@material-ui/core";
import { WindowSizeContext, WindowSize } from "./windowSizeContext";

interface AppBarProps {
  /** Displayed text for the appbar */
  title: string;
  /** An optional function to handle 'go back' action. If defined, a back button will be shown. */
  onBack?: () => void;
  /** Optional nodes to add between title and resize buttons. */
  children?: ReactElement;
}

/** A unified header for all pages within the chathead. */
export const Appbar = ({title, onBack, children = null}: AppBarProps) => (
  <AppBar position="static">
    <Toolbar>
      {
        onBack && (
          <IconButton edge="start" color="inherit" onClick={onBack}>
            <ArrowBackIcon/>
          </IconButton>
        )
      }
      <Typography variant="h6">{title}</Typography>
      <Box flexGrow={1}/>
      {children}
      <WindowSizeContext.Consumer>
        {({size, setSize}) => (
          <>
            <IconButton id="size-collapse" color="inherit" onClick={() => setSize(WindowSize.COLLAPSED)}>
              <PhotoSizeSelectSmallIcon/>
            </IconButton>
            {
              size !== WindowSize.FULL_SCREEN ?
                <IconButton id="size-fullscreen" color="inherit" onClick={() => setSize(WindowSize.FULL_SCREEN)}>
                  <SettingsOverscanIcon/>
                </IconButton>
                :
                <IconButton id="size-regular" color="inherit" onClick={() => setSize(WindowSize.REGULAR)}>
                  <PictureInPictureIcon/>
              </IconButton>
            }
          </>
        )}
      </WindowSizeContext.Consumer>
    </Toolbar>
  </AppBar>
)