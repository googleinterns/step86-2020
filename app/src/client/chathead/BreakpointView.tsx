import React from "react";
import { Accordion, AccordionSummary, Typography, List, ListItem, ListItemText, AccordionDetails, CircularProgress } from "@material-ui/core";
import { Alert, AlertTitle } from "@material-ui/lab";
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ErrorIcon from '@material-ui/icons/Error';
import { Variable } from "../../common/types/debugger";

/** Used to display a breakpoint that has not yet hit. */
export const PendingBreakpointView = ({ breakpointMeta }) => {
  const {location} = breakpointMeta;
  return (
    <Accordion>
      <AccordionSummary disabled expandIcon={<CircularProgress/>}>
        <Typography>{location.path}:{location.line}</Typography>
      </AccordionSummary>
    </Accordion>
  );
};

/** Used to display data for a breakpoint that has already hit. */
export const CompletedBreakpointView = ({ breakpoint }) => {
  console.log(breakpoint);
  const {status} = breakpoint;
  if (status && status.isError) {
    return <FailedCompletedBreakpointView breakpoint={breakpoint}/>;
  }
  return <SuccessfulCompletedBreakpointView breakpoint={breakpoint}/>;
};

/** Shows stackframe data for a successful breakpoint. */
export const SuccessfulCompletedBreakpointView = ({breakpoint}) => {
  const {stackFrames, location} = breakpoint;
  const stackframe = stackFrames[0];
  return (
    <Accordion defaultExpanded>
      <AccordionSummary expandIcon={<ExpandMoreIcon/>}>
        <Typography>{location.path}:{location.line}</Typography>
      </AccordionSummary>
      <AccordionDetails>
        <VariablesView variables={stackframe.locals}/>
      </AccordionDetails>
    </Accordion>
  );
}

/** Shows error data for a failed breakpoint. */
export const FailedCompletedBreakpointView = ({breakpoint}) => {
  const {location, status} = breakpoint;
  // Debugger returns the error message in the form of a bash style template string.
  // Example: format = "Hello $0", parameters = ["Bob"]
  const {format, parameters} = status.description;
  // This Regex looks for sequences like $0, $1, ... and replaces them with the parameter for their index.
  const message = format.replace(/\$({\d}+)/, (match, index) => parameters[index]);
  return (
    <Accordion defaultExpanded>
      <AccordionSummary expandIcon={<ErrorIcon/>}>
        <Typography>{location.path}:{location.line}</Typography>
      </AccordionSummary>
      <AccordionDetails>
        <Alert severity="error">
          {/* Currently this title causes issues with the width of the chathead */}
          {/* <AlertTitle>{status.refersTo}</AlertTitle> */}
          {message}
        </Alert>
      </AccordionDetails>
    </Accordion>
  );
}

/** Displays a set of variables from a debugger stack frame. */
const VariablesView = ({variables}: {variables: Variable[]}) => {
  return (
    <List dense>
      {
        variables.map(variable => (
          <ListItem>
            <ListItemText primary={variable.name} secondary={variable.value}/>
          </ListItem>
        ))
      }
    </List>
  )
}