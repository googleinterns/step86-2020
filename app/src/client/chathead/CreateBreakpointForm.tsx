import React from "react";
import { TextField, Card, CardContent, Button, Box, List, ListItem, ListItemSecondaryAction, IconButton, ListItemText, OutlinedInput, InputAdornment, AccordionSummary, Typography, AccordionDetails, Accordion } from "@material-ui/core";
import DeleteIcon from "@material-ui/icons/Delete";

interface CreateBreakpointFormProps {
  createBreakpoint: (fileName: string, lineNumber: number) => void;
}

interface CreateBreakpointFormState {
  fileName: string;
  lineNumber: number;
  condition: string;
  expressions: string[];
}

export class CreateBreakpointForm extends React.Component<
  CreateBreakpointFormProps,
  CreateBreakpointFormState
> {
  constructor(props: CreateBreakpointFormProps) {
    super(props);
    this.state = {
      fileName: undefined,
      lineNumber: undefined,
      condition: "",
      expressions: []
    };
  }

  onFileName(fileName) {
    this.setState({ fileName });
  }

  onLineNumber(lineNumber) {
    this.setState({ lineNumber });
  }

  onCreateBreakpoint() {
    this.props.createBreakpoint(this.state.fileName, this.state.lineNumber);
  }

  render() {
    const { fileName, lineNumber, condition, expressions } = this.state;
    return (
      <Box m={1}>
        <Card elevation={1}>
          <CardContent>
            <form>
              <TextField
                label="File Name"
                style={{width: "100%"}}
                data-testid="fileName"
                value={fileName}
                onChange={(e) => this.onFileName(e.target.value)}
                variant="outlined"
              />
              <br/><br/>
              <TextField
                label="Line Number"
                style={{width: "100%"}}
                data-testid="lineNumber"
                value={lineNumber}
                onChange={(e) => this.onLineNumber(e.target.value)}
                variant="outlined"
              />
              <br/><br/>
              <ConditionAndExpressionsForm
                condition={condition}
                expressions={expressions}
                setCondition={condition => this.setState({condition})}
                setExpressions={expressions => this.setState({expressions})}
              />
              <br/><br/>
              <Button onClick={(e) => {
                e.preventDefault(); // Prevents a page reload from form submit.
                this.onCreateBreakpoint();
                }}>
                Create Breakpoint
              </Button>
            </form>
          </CardContent>
        </Card>
      </Box>      
    );
  }
}

const ConditionAndExpressionsForm = ({condition, expressions, setCondition, setExpressions}) => {
  return (
    <Accordion>
      <AccordionSummary>
        <Typography variant="body2">Condition and Expressions</Typography>
      </AccordionSummary>
      <AccordionDetails>
        <List>
          <TextField
            label="Condition"
            size="small"
            style={{width: "100%"}}
            variant="outlined"
            value={condition}
            onChange={e => setCondition(e.target.value)}
          />
          <List>
            {
              expressions.map((expression, index) => (
                <ListItem>
                  <TextField
                    size="small"
                    label={`Expression ${index + 1}`}
                    variant="outlined"
                    value={expression}
                    InputProps={{
                      endAdornment: (
                        <InputAdornment position="end">
                          <IconButton onClick={() => {
                            const updatedExpressions = [...expressions];
                            updatedExpressions.splice(index, 1);
                            setExpressions(updatedExpressions);
                          }}>
                            <DeleteIcon/>
                          </IconButton>
                        </InputAdornment>
                      )
                    }}
                  />
                </ListItem>
              ))
            }
            <Button size="small" onClick={() => setExpressions([...expressions, ""])}>Add Expression</Button>
          </List>
        </List>
      </AccordionDetails>
    </Accordion>
  )
}