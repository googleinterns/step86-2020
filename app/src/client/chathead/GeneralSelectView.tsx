import React from "react";
import { Select, MenuItem, FormControl, InputLabel, CircularProgress, TextField } from "@material-ui/core";
import Autocomplete from '@material-ui/lab/Autocomplete';
interface SelectViewProps {
  options: any[];
  label?: string;
  optionsLoading: boolean;
  selectedOptionId?: string;
  // Gets an ID to internally represent an option (will be stored and sent to API)
  optionToId: (option: any) => string;
  optionToLabel: (option: any) => string;
  onChange: (optionId: string) => void;
}

export class SelectView extends React.Component<SelectViewProps, {}> {
  onChange(optionId) {
    console.log(optionId);
    this.props.onChange(optionId);
  }

  render() {
    const { options, selectedOptionId, optionsLoading, optionToId, label, optionToLabel }  = this.props;

    if (optionsLoading) {
      return <CircularProgress/>
    }

    return (
      <FormControl variant="outlined" style={{width: "100%"}}>
        {/* <InputLabel>{label}</InputLabel> */}
        <Autocomplete
          options={options}
          fullWidth
          getOptionLabel={optionToLabel}
          getOptionSelected={(option, value) => optionToId(option) === value}
          value={selectedOptionId}
          onChange={(event, newValue) => this.onChange(optionToId(newValue))}
          renderInput={(params) => <TextField {...params} label={label} variant="outlined"/>}
        />
      </FormControl>
    );
  }
}