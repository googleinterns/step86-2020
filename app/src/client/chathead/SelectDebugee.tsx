import React from "react";

interface SelectDebugeeContainerProps {
  projectId: string;
  debugeeId?: string;
  loadDebugees: () => Promise<any[]>;
  onChange: (debugeeId) => void;
}

interface SelectDebugeeContainerState {
  /** All projects this user has access to, loaded using FetchProjectsRequest */
  debugees: any[]; // TODO: Create a Project type
  debugeesLoading: boolean;
}

export class SelectDebugeeContainer extends React.Component<
  SelectDebugeeContainerProps,
  SelectDebugeeContainerState
> {
  constructor(props: SelectDebugeeContainerProps) {
    super(props);

    this.state = {
      debugees: [],
      debugeesLoading: false,
    };
  }

  onChange(debugeeId) {
    this.props.onChange(debugeeId);
  }

  componentDidMount() {
    this.setState({ debugeesLoading: true });
    this.props.loadDebugees().then((debugees) => {
      this.setState({ debugees, debugeesLoading: false });
    });
  }

  render() {
    return (
      <div>
         <span>Project: {this.props.projectId}</span>
         <SelectDebugeeView
        debugees={this.state.debugees}
        debugeesLoading={this.state.debugeesLoading}
        debugeeId={this.props.debugeeId}
        onChange={(debugeeId) => this.onChange(debugeeId)}
      />
      </div>

    );
  }
}

interface SelectDebugeeViewProps {
  debugees: any[];
  debugeesLoading: boolean;
  debugeeId?: string;
  onChange: (debugeeId: string) => void;
}

export class SelectDebugeeView extends React.Component<
  SelectDebugeeViewProps,
  {}
> {
  onChange(debugeeId) {
    this.props.onChange(debugeeId);
  }

  render() {
    const { debugees, debugeeId, debugeesLoading } = this.props;
    return (
      <div>
        <h1>Select Debugee</h1>
        {debugeesLoading === true && <LoadingView />}
        {debugeesLoading === false && (
          <DebugeeSelect
            {...{
              debugeeId,
              debugees,
              onChange: (debugeeId) => this.onChange(debugeeId),
            }}
          />
        )}
      </div>
    );
  }
}

export const LoadingView = () => <div>Loading...</div>;
export const DebugeeSelect = ({ debugees, debugeeId, onChange }) => {
  return (
    <select value={debugeeId} onChange={(e) => onChange(e.target.value)}>
      {debugees.map((debugee) => (
        <DebugeeOption key={debugee} debugee={debugee} />
      ))}
    </select>
  );
};

export const DebugeeOption = ({ debugee }) => (
  <option value={debugee}>{debugee}</option>
);
