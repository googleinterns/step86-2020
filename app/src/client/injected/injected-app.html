<!DOCTYPE html>

<html>
  <head>
    <script src="./injected-app.tsx"></script>
  </head>
  <body>
    <div id="injected-app" />
  </body>
</html>
