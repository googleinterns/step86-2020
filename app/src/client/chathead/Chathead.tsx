import React from "react";
import styled from "styled-components";
import { SelectProjectContainer } from "./SelectProject";
import { SelectDebuggeeContainer } from "./SelectDebuggee";
import { CreateBreakpointForm } from "./CreateBreakpointForm";
import {
  FetchProjectsRequest,
  FetchProjectsRequestData,
  FetchDebuggeesRequestData,
  FetchDebuggeesRequest,
} from "../../common/requests/BackgroundRequest";
import { BreakpointMeta, Breakpoint } from "../../common/types/debugger";
import {
  PendingBreakpointView,
  CompletedBreakpointView,
} from "./BreakpointView";

import Paper from "@material-ui/core/Paper";
import { AppBar, Toolbar, Typography, IconButton } from "@material-ui/core";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
// Jest has trouble with this, so only import in actual builds.
if (process.env.NODE_ENV !== "test") {
  import("fontsource-roboto");
}
interface ChatheadProps {
  /** The current selected project. Undefined if no project selected. */
  projectId: string | undefined;
  /** The current selected debuggee. Undefined if no debuggee selected. */
  debuggeeId: string | undefined;
  /** Breakpoints that have been set but not hit. */
  activeBreakpoints: BreakpointMeta[];
  /** Breakpoints that have been hit, with all data loaded. */
  completedBreakpoints: Breakpoint[];

  setProject: (projectId: string) => void;
  setDebuggee: (debuggeeId: string) => void;
  createBreakpoint: (fileName: string, lineNumber: number) => void;
  deleteBreakpoint: (breakpointId: string) => void;
  deleteAllActiveBreakpoints: () => void;
}

interface ChatheadState {}

export class Chathead extends React.Component<ChatheadProps, ChatheadState> {
  constructor(props: ChatheadProps) {
    super(props);
  }

  render() {
    const { projectId, debuggeeId, projectDescription } = this.props;
    return (
      <ChatheadWrapper>
        {!projectId && (
          <SelectProjectContainer
            projectId={this.props.projectId}
            projectDescription={this.props.projectDescription}
            onChange={this.props.setProject}
            loadProjects={async () => {
              const response = await new FetchProjectsRequest().run(
                new FetchProjectsRequestData()
              );
              return response.projects;
            }}
          />
        )}

        {projectId && !debuggeeId && (
          <SelectDebuggeeContainer
            projectId={this.props.projectId}
            debuggeeId={this.props.debuggeeId}
            onChange={this.props.setDebuggee}
            loadDebuggees={async () => {
              const response = await new FetchDebuggeesRequest().run(
                new FetchDebuggeesRequestData(this.props.projectId)
              );
              // Response.debuggees will be undefined if there are no active debuggees.
              return response.debuggees || [];
            }}
            backToProjects={() => {
              this.props.setProject(undefined);
            }}
          />
        )}

        {projectId && debuggeeId && (
          <>
            <AppBar position="static">
              <Toolbar>
                <IconButton
                  edge="start"
                  color="inherit"
                  onClick={() => this.props.setDebuggee(undefined)}
                >
                  <ArrowBackIcon />
                </IconButton>
                <Typography variant="h6">Breakpoints</Typography>
              </Toolbar>
            </AppBar>
            <CreateBreakpointForm
              createBreakpoint={this.props.createBreakpoint}
              deleteAllActiveBreakpoints={this.props.deleteAllActiveBreakpoints}
            />
          </>
        )}

        {this.props.activeBreakpoints.map((b) => (
          <PendingBreakpointView breakpointMeta={b} />
        ))}

        {this.props.completedBreakpoints.map((b) => (
          <CompletedBreakpointView
            breakpoint={b}
            deleteBreakpoint={this.props.deleteBreakpoint}
          />
        ))}
      </ChatheadWrapper>
    );
  }
}

const ChatheadWrapper = styled(Paper)`
  position: fixed;
  top: 20px;

  right: 20px;
  width: fit-content;

  z-index: 1000;
`;
