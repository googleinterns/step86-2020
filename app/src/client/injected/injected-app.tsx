import React from "react";
import ReactDOM from "react-dom";
import { Chathead } from "../chathead/Chathead";
import * as BackgroundRequest from "../../common/requests/BackgroundRequest";
import { BreakpointMeta, Breakpoint } from "../../common/types/debugger";
import { BreakpointMarkers } from "../markers/BreakpointMarkers";

interface InjectedAppState {
  projectId: string;
  debuggeeId: string;

  activeBreakpoints: { [key: string]: BreakpointMeta };
  completedBreakpoints: { [key: string]: Breakpoint };

  localStorage?: Storage
}

export class InjectedApp extends React.Component<any, InjectedAppState> {
  constructor(props: InjectedAppState) {
    super(props);
    this.state = {
      projectId: this.getGcpProjectId(),
      debuggeeId: undefined,

      activeBreakpoints: {},
      completedBreakpoints: {},
    };
  }

  /**
   * This function fetches projects names on Github and return the title of the project
   */
  getProjectNameFromGithub(): string {
    let title = document.querySelector(".js-path-segment:first-child");
    if (title !== null) {
      var projectName = title.innerText;
      return projectName;
    }
  }

  /**
   * This function checks of the project name already exists in the local storage
   */
  getGcpProjectId(): string {
    let gcpProjectId = localStorage.getItem(this.getProjectNameFromGithub());
    return gcpProjectId !== null ? gcpProjectId : undefined;
  }

  /** Saves the current project name - project id association (if possible) */
  saveGcpProjectId(projectId: string): void {
    const projectName = this.getProjectNameFromGithub();
    // Only save if we're able to pull the project name.
    if (projectName) {
      // Save new project ID only if there is a projectId.
      // Otherwise, localStorage saves the string "undefined" which throws things off.
      if (projectId) {
        this.getLocalStorage().setItem(projectName, projectId);
      } else {
        // In the undefined case, manually remove the saved project.
        this.getLocalStorage().removeItem(projectName);
      }
    }
  }

  /**
   * Makes request to the BackgroundRequest to set the breakpoint
   * using debuggee id, file name and line number
   * @param {String} filename Name of the file to set the breakpoint in that file.
   * @param {number} lineNumber line number to set the breakpoint on that line.
   */
  async createBreakPoint(fileName: string, lineNumber: number, condition: string = "", expressions: string[] = []) {
    // Make the Set breakpoint request in the BackgroundRequest
    const response = await new BackgroundRequest.SetBreakpointRequest().run(
      new BackgroundRequest.SetBreakpointRequestData(
        this.state.debuggeeId,
        fileName,
        lineNumber,
        condition,
        expressions
      )
    );
    /**
     * Side note: you cannot mutate the state variable in React. So, to append (add) values to an array, you have to
     * make a temporary copy of state array and push elements to it. And then reassign the state variable
     */
    // Add the response to the state object variable
    var tempActiveBreakpoint = { ...this.state.activeBreakpoints };
    tempActiveBreakpoint[response.breakpoint.id] = response.breakpoint;
    this.setState({ activeBreakpoints: tempActiveBreakpoint });
  }

  /**
   * This function is invoked immediately after a component is mounted. This allows to
   * keep calling listbreakpoints request on an interval of 5 seconds to check
   * the difference between active breakpoints and non-active breakpoints.
   */
  componentDidMount() {
    // Set waitToken to null as default for first call
    var waitToken = null;
    // Checks if debuggeeId is not undefined
    setInterval(async () => {
      if (
        this.state.debuggeeId !== undefined &&
        Object.keys(this.state.activeBreakpoints).length > 0
      ) {
        // Make the list breakpoint request
        let breakpointListResponse = await new BackgroundRequest.ListBreakPointsRequest().run(
          new BackgroundRequest.ListBreakpointsData(
            this.state.debuggeeId,
            waitToken
          )
        );
        waitToken = breakpointListResponse.nextWaitToken;

        let breakpointList: Array<any> = [];
        // Add the active breakpoints to the listBreakpoint state array.
        for (let breakpoint of breakpointListResponse.breakpoints || []) {
          breakpointList.push(breakpoint["id"]);
        }

        // Get the inactive breakpoints (by getting difference from activeBreakpoint and breakpointList) array
        let inactiveBreakpoints: Array<any> = [];
        for (let breakpointId in this.state.activeBreakpoints) {
          if (!breakpointList.includes(breakpointId)) {
            inactiveBreakpoints.push(breakpointId);
          }
        }

        // Call function to get the breakpoint data sending non-active breakpoints.
        this.loadBreakpoints(inactiveBreakpoints);
      }
    }, 5000);
  }

  /** Deletes a breakpoint from debugger backend. */
  async deleteBreakpoint(breakpointId: string) {
    // Delete the breakpoint from remote cloud debugger.
    const deleteBreakpointRequest = await new BackgroundRequest.DeleteBreakpointRequest().run(
      new BackgroundRequest.DeleteBreakpointRequestData(
        this.state.debuggeeId,
        breakpointId
      )
    );
    // Remove breakpoint from local tracking.
    const updatedCompletedBreakpoints = { ...this.state.completedBreakpoints };
    delete updatedCompletedBreakpoints[breakpointId];
    this.setState({ completedBreakpoints: updatedCompletedBreakpoints });
  }

  /**
   * This function gets the data of non-active breakpoints (breakpoint that are hit)
   * and saves it to the getBreakpoint state array. Moreover removes it from active breakpoint array.
   * @param {Array<any>} differenceBreakpoint list of non-active breakpoints.
   */
  async loadBreakpoints(breakpointIdsToLoad: Array<any>) {
    // Make request to get the breakpoint data using breakpoint ids
    var tempGetBreakpoints = { ...this.state.completedBreakpoints };
    var updatedActiveBPs = { ...this.state.activeBreakpoints };
    for (let breakpointId of breakpointIdsToLoad) {
      const getBreakpointresponse = await new BackgroundRequest.FetchBreakpointRequest().run(
        new BackgroundRequest.FetchBreakpointRequestData(
          this.state.debuggeeId,
          breakpointId
        )
      );
      // Add it to the state array for getBreakpoint data
      const { breakpoint } = getBreakpointresponse;
      tempGetBreakpoints[breakpoint.id] = breakpoint;
      // Remove the breakpoint from the active breakpoint list
      delete updatedActiveBPs[breakpointId];
    }
    this.setState({ completedBreakpoints: tempGetBreakpoints });
    this.setState({ activeBreakpoints: updatedActiveBPs });
  }
  /**
   * Delete a batch of active breakpoints from debugger backend
   */
  async deleteAllActiveBreakpoints() {
    //delete all active breakpoints from remote cloud debugger.
    for (let breakpointId of Object.values(this.state.activeBreakpoints)) {
      const deletionRequest = await new BackgroundRequest.DeleteBreakpointRequest().run(
        new BackgroundRequest.DeleteBreakpointRequestData(
          this.state.debuggeeId,
          breakpointId.id
        )
      );
    }
    this.setState({ completedBreakpoints: {} });
    this.setState({ activeBreakpoints: {} });
  }

  render() {
    /** Active and completed breakpoints are kept in state differently because active breakpoints must be deletable.
     *  However, as far as UI is concerned, they are both lists. This conversion helps simplify markup.
     */
    const activeBreakpoints = Object.values(this.state.activeBreakpoints);
    const completedBreakpoints = Object.values(this.state.completedBreakpoints);

    return (
      <>
        <Chathead
          projectId={this.state.projectId}
          debuggeeId={this.state.debuggeeId}
          activeBreakpoints={activeBreakpoints}
          completedBreakpoints={completedBreakpoints}
          setProject={(projectId) => {
            localStorage.setItem(this.getProjectNameFromGithub(), projectId);
            this.setState({ projectId });
          }}
          setDebuggee={(debuggeeId) => this.setState({ debuggeeId })}
          createBreakpoint={(fileName, lineNumber) =>
            this.createBreakPoint(fileName, lineNumber)
          }
          deleteBreakpoint={(breakpointId: string) =>
            this.deleteBreakpoint(breakpointId)
          }
          deleteAllActiveBreakpoints={() => this.deleteAllActiveBreakpoints()}
        />

        <BreakpointMarkers
          activeBreakpoints={activeBreakpoints}
          completedBreakpoints={completedBreakpoints}
          createBreakpoint={(fileName, lineNumber) =>
            this.createBreakPoint(fileName, lineNumber)
          }
        />
      </>
    );
  }

  /** Get the windows local storage object. This accepts a prop to allow mocks. */
  getLocalStorage() {
    return this.props.localStorage || window.localStorage;
  }
}

// Mounts injected App once html is loaded
const mount = document.createElement("div");
document.body.appendChild(mount);
ReactDOM.render(<InjectedApp />, mount);
