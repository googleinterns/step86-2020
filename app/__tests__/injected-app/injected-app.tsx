import React from "react";
import { shallow, mount, render, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

import { InjectedApp } from "../../src/client/injected/injected-app";
import { Chathead } from "../../src/client/chathead/Chathead";

describe("Injected App", () => {
    it("displays injected app when no project selected.", () => {
      const wrapper = mount(
        <InjectedApp
          projectId={undefined}
          debuggeeId={undefined}
          breakpoints={[]}
          lineNum={5}
          fileName={"index.js"}
          activeBreakpoints={[]}
          completedBreakpoints={[]}
        />
      );
      expect(wrapper.find(Chathead)).toHaveLength(1);
    });

    it("checks getGcpProjectId method is called", () => {
      jest.spyOn(InjectedApp.prototype, 'getGcpProjectId');
      shallow(<InjectedApp />);
      expect(InjectedApp.prototype.getGcpProjectId).toHaveBeenCalled();
    });

    it("displays injected app when no project selected.", () => {
      jest.spyOn(InjectedApp.prototype, 'getProjectNameFromGithub');
      shallow(<InjectedApp />);
      expect(InjectedApp.prototype.getProjectNameFromGithub).toHaveBeenCalled();
    });
});
  