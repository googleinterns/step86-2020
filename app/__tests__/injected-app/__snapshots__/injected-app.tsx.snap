// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Injected App displays injected app when no project selected. 1`] = `
Array [
  <React.Fragment>
    <Chathead
      activeBreakpoints={Array []}
      completedBreakpoints={Array []}
      createBreakpoint={[Function]}
      setDebuggee={[Function]}
      setProject={[Function]}
    />
  </React.Fragment>,
]
`;
